File: python/ch01_agent/agent_ch01/bootstrap.py
"""组合根：把外部实现装配为第 1 章 Agent。

Java 对照：可以把本文件理解为 Spring 的 `@Configuration` 类。
它负责创建对象和连接依赖，但不负责真正的 Agent 业务流程。
"""

from .adapters.powershell import PowerShellRunner
from .core.commands import CommandRunner
from .core.loop import AgentRunner, ToolAuthorizer
from .core.model import ModelClient
from .core.profiles import ChapterProfile
from .features.builtin_tools import create_chapter_one_tools

SYSTEM_PROMPT = "You are a coding agent. Use tools when needed, inspect their results, and answer accurately."


def build_agent(profile: ChapterProfile, model: ModelClient, workspace: str, command_runner: CommandRunner | None = None, authorizer: ToolAuthorizer | None = None, max_turns: int = 20) -> AgentRunner:
    """创建一个第 1 章 AgentRunner。

    `command_runner or PowerShellRunner()` 表示：调用者传了 Fake 就用 Fake，
    没传时才创建真实 PowerShellRunner。这就是最简单的依赖注入。
    """
    if profile.chapter != 1:
        raise ValueError(f"Chapter {profile.chapter} has not been migrated to Python yet")
    return AgentRunner(model, create_chapter_one_tools(command_runner or PowerShellRunner()), SYSTEM_PROMPT, workspace, max_turns=max_turns, authorizer=authorizer)

File: typescript/chapters/ch01/src/bootstrap.ts
import { PowerShellRunner } from "./adapters/powershell.js";
import type { CommandRunner } from "./core/commands.js";
import { AgentRunner } from "./core/loop.js";
import type { ToolAuthorizer } from "./core/loop.js";
import type { ModelClient } from "./core/model.js";
import type { ChapterProfile } from "./core/profiles.js";
import { createChapterOneTools } from "./features/builtin-tools.js";

// 组合根（Composition Root）：负责把模型适配器、命令适配器和工具注册表装配为 AgentRunner。
//
// 核心循环只依赖 ModelClient / CommandRunner / ToolRegistry 等接口，
// 具体进程与 SDK 的创建集中在这里，测试也能注入替身隔离网络和操作系统。
// buildAgent 同时以 ChapterProfile 限制能力白名单，防止后续章节能力泄漏进第 1 章。
//
// 固定系统提示词属于组装配置；循环在每轮模型请求前注入它。
const SYSTEM_PROMPT =
  "You are a coding agent. Use tools when needed, inspect their results, and answer accurately.";

// 组装层只负责注入章节依赖，核心循环不直接依赖具体进程实现。
export interface BuildDependencies {
  // 调用方提供模型边界，测试可注入确定性替身以隔离网络。
  readonly model: ModelClient;
  // 绝对或相对工作目录最终由 AgentRunner 规范化并传给工具。
  readonly workspace: string;
  // 未注入时才创建真实 PowerShell 适配器，避免测试依赖进程。
  readonly commandRunner?: CommandRunner;
  // 可选人工授权边界；未提供时按本章最小循环直接执行工具。
  readonly authorizer?: ToolAuthorizer;
  // 为循环设置有限回合预算，防止模型持续请求工具。
  readonly maxTurns?: number;
}

// 组合根：把第 1 章允许的工具、模型和运行上下文装配为单一循环实例。
// profile 必须属于本章节；dependencies 只承载基础设施而不承载业务流程。
export function buildAgent(profile: ChapterProfile, dependencies: BuildDependencies): AgentRunner {
  // 固定章节入口拒绝更高版本 profile，防止能力从后续章节意外泄漏。
  if (profile.chapter !== 1) {
    throw new Error(`Chapter ${profile.chapter} has not been migrated to TypeScript yet`);
  }
  // 仅在依赖未注入时使用默认适配器，其他依赖原样透传给核心循环。
  return new AgentRunner({
    model: dependencies.model,
    tools: createChapterOneTools(
      dependencies.commandRunner === undefined
        ? new PowerShellRunner()
        : dependencies.commandRunner,
    ),
    systemPrompt: SYSTEM_PROMPT,
    workspace: dependencies.workspace,
    ...(dependencies.authorizer === undefined ? {} : { authorizer: dependencies.authorizer }),
    ...(dependencies.maxTurns === undefined ? {} : { maxTurns: dependencies.maxTurns }),
  });
}
